kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt

```diff
@@ -0,0 +1,3087 @@
+                "bufferEndSegment should not have lower id than receiveSegment.\n" +
+                    "Channel state: $this"
+            }
+        }
+        val firstSegment = listOf(receiveSegment.value, sendSegment.value, bufferEndSegment.value)
+            .filter { it !== NULL_SEGMENT }
+            .minBy { it.id }
+        check(firstSegment.prev == null) {
+            "All processed segments should be unreachable from the data structure, but the `prev` link of the leftmost segment is non-null.\n" +
+                "Channel state: $this"
+        }
+        // Check that the doubly-linked list of segments does not
+        // contain full-of-cancelled-cells segments.
+        var segment = firstSegment
+        while (segment.next != null) {
+            // Note that the `prev` reference can be `null` if this channel is closed.
+            check(segment.next!!.prev == null || segment.next!!.prev === segment) {
+                "The `segment.next.prev === segment` invariant is violated.\n" +
+                    "Channel state: $this"
+            }
+            // Count the number of closed/interrupted cells
+            // and check that all cells are in expected states.
+            var interruptedOrClosedCells = 0
+            for (i in 0 until SEGMENT_SIZE) {
+                when (val state = segment.getState(i)) {
+                    BUFFERED -> {} // The cell stores a buffered element.
+                    is Waiter -> {} // The cell stores a suspended request.
+                    INTERRUPTED_RCV, INTERRUPTED_SEND, CHANNEL_CLOSED -> {
+                        // The cell stored an interrupted request or indicates
+                        // that this channel is already closed.
+                        // Check that the element slot is cleaned and increment
+                        // the number of cells in closed/interrupted state.
+                        check(segment.getElement(i) == null)
+                        interruptedOrClosedCells++
+                    }
+                    POISONED, DONE_RCV -> {
+                        // The cell is successfully processed or poisoned.
+                        // Check that the element slot is cleaned.
+                        check(segment.getElement(i) == null)
+                    }
+                    // Other states are illegal after all running operations finish.
+                    else -> error("Unexpected segment cell state: $state.\nChannel state: $this")
+                }
+            }
+            // Is this segment full of cancelled/closed cells?
+            // If so, this segment should be removed from the
+            // linked list if nether `receiveSegment`, nor
+            // `sendSegment`, nor `bufferEndSegment` reference it.
+            if (interruptedOrClosedCells == SEGMENT_SIZE) {
+                check(segment === receiveSegment.value || segment === sendSegment.value || segment === bufferEndSegment.value) {
```

Comment:
This is a special internal method that checks that the data structure is correct. It must throw an exception otherwise. While `assert`s are always on in tests, I would prefer not to rely on it, making the method more robust.